### PR TITLE
Buffer struct

### PR DESCRIPTION
**Pull request by bdevierno1**
<<Replace this line with a short description of the changes>>

## Summary:

**Usage:**

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        |
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [ ] PR is ready for review

## Review: 
(optional) << @-mention people who should review these changes >>

(optional) **Subscribers:** << @-mention people who probably care about these changes >>

### CODE REVIEW

File: examples/api_gateway/api_gateway.h [hunk 1]
```diff
@@ -166,6 +169,9 @@ struct onvm_parser_ipv4_5tuple {
  */
 struct data {
```
Comment (kevindweb) [inline examples/api_gateway/api_gateway.h:169]: Can you add the rx_fd to this struct and assign it in `dequeue_and_free_buffer_map`?

File: examples/api_gateway/scaler.c [hunk 2]
```diff
@@ -212,19 +204,11 @@ move_buffer_to_container(void) {
                  * Add rx pipe to epoll for polling
                  */
 
-                // TODO: @bdevierno1 need to modify this dequeue function to push buffered packets to tx_pipe
-                // push all buffered packets to the pipe
-                if (dequeue_and_free_buffer_map(flow, ring, pipe->tx_pipe) < 0) {
+                int rx_fd = 0; // TODO: get the corresponding rx_fd number.
```
Comment (bdevierno1) [inline examples/api_gateway/scaler.c]: Added a todo here. We plan to do this later correct?

File: examples/api_gateway/scaler.c [hunk 3]
```diff
@@ -212,19 +204,11 @@ move_buffer_to_container(void) {
...
+                if (dequeue_and_free_buffer_map(flow, pipe->tx_pipe, rx_fd) < 0) {
```
Comment (kevindweb) [inline examples/api_gateway/scaler.c]: Probably want it to be this.
```suggestion
                if (dequeue_and_free_buffer_map(flow, pipe->tx_pipe, pipe->rx_pipe) < 0) {
```

File: examples/api_gateway/scaler.c [hunk 4]
```diff
@@ -212,19 +204,11 @@ move_buffer_to_container(void) {
...
                         continue;
                 }
-
-                struct data *data = NULL;
-                if (onvm_ft_add_key(em_tbl, flow, (char **)&data) < 0) {
-                        perror("Couldn't add IP flow to flow table");
-                        continue;
-                }
-
                 // TODO: @bdevierno1 need to keep this to give polling thread more pipes
```
Comment (kevindweb) [inline examples/api_gateway/scaler.c]: This too
```suggestion
```

File: examples/api_gateway/api_gateway.c [hunk 5]
```diff
@@ -271,6 +274,9 @@ init_rings(void) {
         const unsigned ring_size = 64;
 
         // TODO: @bdevierno1 need to add a ring of onvm_ft_ipv4_5tuple here
```
Comment (kevindweb) [inline examples/api_gateway/api_gateway.c]: Just want to clean up these lines.
```suggestion
```

File: examples/api_gateway/scaler.c (hunk 2 above)
Comment (kevindweb) [inline examples/api_gateway/scaler.c]: Don't need this, see the suggestion below.
```suggestion
```

File: examples/api_gateway/api_gateway.h [hunk 6]
```diff
@@ -43,13 +43,16 @@
...
 #include "onvm_pkt_common.h"
 
+#include <rte_rwlock.h>
+
 #define NUM_CONTAINERS 4
 #define CONT_RX_PIPE_NAME "/tmp/rx/%d"
 #define CONT_TX_PIPE_NAME "/tmp/tx/%d"
 #define PKTMBUF_POOL_NAME "MProc_pktmbuf_pool"
 #define FLOW_RING_NAME "IPv4_Flow_%u_%d"
 #define _GATE_2_BUFFER "GATEWAY_2_BUFFER"
```
Comment (kevindweb) [inline examples/api_gateway/api_gateway.h]: While we're cleaning up, can we start by getting rid of unused rings?
```suggestion
```

File: examples/api_gateway/api_gateway.h [hunk 7]
```diff
@@ -83,7 +86,7 @@ rte_atomic16_t num_running_containers;
 
 // buffer pulls from gateway and scaler ring buffers
 
-struct rte_ring *scale_buffer_ring, *gate_buffer_ring;
+struct rte_ring *scale_buffer_ring, *gate_buffer_ring, *container_init_ring;
```
Comment (kevindweb) [inline examples/api_gateway/api_gateway.h]: Gateway Buffer ring may not be useful.
```suggestion
struct rte_ring *scale_buffer_ring, *container_init_ring;
```

File: examples/api_gateway/api_gateway.c [hunk 8]
```diff
@@ -271,6 +274,9 @@ init_rings(void) {
...
+        container_init_ring = rte_ring_create(_INIT_CONT_TRACKER, ring_size, rte_socket_id(), flags);
+        if (container_init_ring == NULL)
+                rte_exit(EXIT_FAILURE, "Problem getting receiving ring\n");
         gate_buffer_ring = rte_ring_create(_GATE_2_BUFFER, ring_size, rte_socket_id(), flags);
         if (gate_buffer_ring == NULL)
                 rte_exit(EXIT_FAILURE, "Problem getting receiving ring\n");
```
Comment (kevindweb) [inline examples/api_gateway/api_gateway.c]: This one doesn't seem used anymore.
```suggestion
```

**Review by kevindweb — CHANGES_REQUESTED** [review]
Sorry for all the suggestions, but I feel like this will help clean up the code. I can do them if you want though.

**Review by kevindweb — APPROVED** [review]
Awesome @bdevierno1 and @catherinemeadows !

**Pull request merged**